sycl/source/detail/plugin.hpp

```diff
@@ -184,11 +184,71 @@ class plugin {
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
 
+  // return the index of PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return It - PiPlatforms->begin();
+
+      PiPlatforms->push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms->size() - 1;
+    }
+
+    return -1;
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
```

Comment:
Can we have a situation when multiple threads are calling these `get/set` functions?

Reply:
`setLastDeviceId` will be called only once when we initialize a plugin.
`getStartingDeviceId` can be called from multiple threads, but they are read-only.

Reply:
```
    if (!PiPlatforms)
      PiPlatforms = new std::vector<RT::PiPlatform>;
```
These lines do look like thread safe. It can happen that two threads pass the  check and both allocate the memory.

Reply:
we use std::call_once for RT::initialize() function.

Reply:
Done.
A lock is added in the only caller filterDeviceFilter() in platform_impl.cpp.